packages/docusaurus/src/client/PendingNavigation.tsx

```diff
@@ -94,6 +94,15 @@ class PendingNavigation extends React.Component<Props, State> {
     return true;
   }
 
+  componentDidUpdate() {
```

Comment:
Note that I don't want to have a subcomponent just to dispatch this lifecycle.

1. The `onRouteUpdateDelayed` has to be dispatched in the `PendingNavigation` (because it's fired within a promise callback), so let's just dispatch everything here to avoid scattered code.
2. The `render()` of this component directly returns a `Route`, while having an extra layer makes the architecture not so elegant anymore...

Reply:
1) there's something important to consider: do we want to fire the event sync or async? useEffect or useLayoutEffect? (componentDidMount is closer to useLayoutEffect)

I'd rather immediately use a hook today because when we'll refactor this component later we won't risk to change this async vs sync behavior (unless you find a way to test this 😉 ). People may rely on this and I'd rather be explicit about it (even in our docs)

https://reactjs.org/docs/hooks-effect.html

<img width="939" alt="image" src="https://user-images.githubusercontent.com/749374/155346769-62de1547-1e06-4029-b985-cdbf7208d885.png">

IMHO by default this event will be quite used by analytics and reporting. For performance reasons I'd rather it to be async, just after the page transition completed, instead of blocking it. Hence using `useEffect` which is not possible to achieve with a class.

There might be value in having a sync lifecycle that blocks the transition but let's see if users come with a need for this first.

Also, something to consider: what happens when the lifecycle callback throws? I believe the behavior wouldn't be the same in both cases. 

Do we want to fail-fast or fail-safe if an analytics plugin has a bug?

2) I believe this should work

```tsx
      <Route
        location={normalizeLocation(location)}
        render={() => (
          <>
            {children}
            <ClientLifecyclesDispatcher />
          </>
        )}
      />
```

Reply:
> Do we want to fail-fast or fail-safe if an analytics plugin has a bug?

It would be up to the user to catch their error. We just fire the lifecycle.

> I believe this should work

Looks a bit too cluttered IMO, that's what I'm trying to avoid... Could this work?

```js
<ClientLifecyclesDispatcher>
  <Route />
< ClientLifecyclesDispatcher />
```

Also, we always have to put `onRouteUpdateDelayed` within the promise callback, which means the lifecycle dispatching will now be scattered within two components.

Reply:
> It would be up to the user to catch their error. We just fire the lifecycle.

If it's up to the user (and plugin authors) then we must document that appropriately: if their lifecycle take a long time or if it can throw they must consider catching errors and execution the code asynchronously

I believe it's better if we do this by default instead of letting the user have to think about this. 

> Could this work?

Not 100% but after taking a look it should also work yes. There's probably no value in using children here:

```tsx
<>
  <ClientLifecyclesDispatcher/>
  <Route />
</>
```

> which means the lifecycle dispatching will now be scattered within two components.

That doesn't seem like a big deal considering this is temporary (we'll refactor this later) and the lifecycles behave exactly how we want them to

Reply:
> if it can throw they must consider catching errors and execution the code asynchronously

The client-side APIs are usually async already. Is there anything that's not callback-/promise-based but render-blocking? Can't think of a compelling use-case.